Core/Scripts: Archmage Arugal event script

tkrokli opened the pull request:
**Changes proposed:**

Implement actual boss script for Archmage Arugal
- add yells used in boss event to DB
- use proper spells instead of only melee fighting
- add DataType in shadowfang_keep.h
- event scheduling in the script

by Riztazz

**Target branch(es):** **3.3.5** (the instance is different in retail and Archmage Arugal is not there)

**Issues addressed:** Closes #17916

**Tests performed:** (Does it build, tested in-game, etc)
- Builds without issues
- Works as intended in-game

**Known issues and TODO list:**
- [x] Check the visual details in-game to see if there is any unwanted behaviour
- [x] Needs feedback from TC developers (and others) for recommended improvements

Commit:
Core/Scripts: Archmage Arugal event script

Implement actual boss script for Archmage Arugal
- add yells used in boss event to DB
- use proper spells instead of only melee fighting
- add DataType in shadowfang_keep.h
- event scheduling in the script

by Riztazz, closes issue 17916.

Riztazz committed:
Core/Scripts: Archmage Arugal event script

Implement actual boss script for Archmage Arugal
- add yells used in boss event to DB
- use proper spells instead of only melee fighting
- add DataType in shadowfang_keep.h
- event scheduling in the script

Closes issue #17916.
Author:    Riztazz

Keader commented on src/server/scripts/EasternKingdoms/ShadowfangKeep/shadowfang_keep.cpp:
```diff
@@ -205,6 +205,131 @@ class npc_arugal_voidwalker : public CreatureScript
+    {
+        boss_archmage_arugalAI(Creature* creature) : BossAI(creature, BOSS_ARUGAL) { }
+
+        void Reset() override
```
realy needed?

tkrokli replied:
Check with Riztazz in #trinity (unless someone else wants to answer this)

Keader commented on src/server/scripts/EasternKingdoms/ShadowfangKeep/shadowfang_keep.cpp:
```diff
@@ -205,6 +205,131 @@ class npc_arugal_voidwalker : public CreatureScript
+
+        void AttackStart(Unit* who) override
+        {
+            if (!who)
```
realy necessary this check? who realy can be null?

tkrokli replied:
Check with Riztazz in #trinity (unless someone else wants to answer this)

Keader commented on src/server/scripts/EasternKingdoms/ShadowfangKeep/shadowfang_keep.cpp:
```diff
@@ -205,6 +205,131 @@ class npc_arugal_voidwalker : public CreatureScript
+
+            events.Update(diff);
+
+            while (auto eventId = events.ExecuteEvent())
```
while (uint32 eventId = events.ExecuteEvent())

Treeston replied:
Yeah, we don't really do `auto`.

tkrokli commented:
```
[2016.09.09, 20:32:51] <Riztazz> const uint32 spellsSize = 3;
[2016.09.09, 20:33:06] <Riztazz> and change this std::array<uint32, 3> arugalTeleportSpells =  to  std::array<uint32, spellsSize> arugalTeleportSpells =
[2016.09.09, 20:33:15] <Riztazz> or somewhere in enum
[2016.09.09, 20:33:38] <Riztazz> and this comment is pointless // rotate array, can be done with some kind of shuffle. Up to you to figure out what is best.
[2016.09.09, 20:33:45] <Riztazz> at this point, it was only useful in issue itself:P
```

Commit:
Recommended script changes & added braces

- added missing braces around arugalTeleportSpells lines
- added const spellsSize instead of "magic number"
- removed void Reset() hook
- removed trailing spaces
- removed if (!who)
- auto -> uint32

Treeston committed:
Code style and logic fixes.

jackpoz commented on src/server/scripts/EasternKingdoms/ShadowfangKeep/shadowfang_keep.cpp:
```diff
@@ -205,6 +204,126 @@ class npc_arugal_voidwalker : public CreatureScript
+                DoMeleeAttackIfReady();
+            }
+
+            uint32 _teleportSpells[NUM_TELEPORT_SPELLS];
```
don't call static and class field with the same name except a small _ difference, it's confusing and prone to errors.
doesn't it make more sense to avoid the whole static array idea and just initialize the class field to whatever you need to ?

tkrokli replied:
Good point, thanks for pointing it out.

jackpoz replied:
what happened to this commend ? was the code updated ? (github is still showing me the static teleportSpells variable)

tkrokli replied:
@Treeston : do you have a suggestion for a change/improvement in this part of the code?

jackpoz commented on src/server/scripts/EasternKingdoms/ShadowfangKeep/shadowfang_keep.cpp:
```diff
@@ -205,6 +205,131 @@ class npc_arugal_voidwalker : public CreatureScript
+                        auto spell = arugalTeleportSpells.front();
+                        DoCast(spell); 
+                        // rotate array, can be done with some kind of shuffle. Up to you to figure out what is best.
+                        std::rotate(arugalTeleportSpells.begin(), arugalTeleportSpells.begin() + 1, arugalTeleportSpells.end());
```
you CANNOT modify a global variable in EVERY class instance of boss_archmage_arugalAI without any locking mechanism, it will cause the very much dreaded race conditions. I strongly suggest you to spend some time reading up about "race condition" to fully understand what and why is bad.

tkrokli replied:
I wonder if Riztazz will have a constructive reply to this / a better solution.
It was after all something he wrote to pass time on a train station.

Treeston replied:
This is an outdated diff. The current version is thread safe assuming only one thread is accessing the creature at any given time (this is the case, each map is single-threaded).

Riztazz replied:
Derp, it was supposed to be within class, dunno why i did it like that. my excuse is it was written quickly while waiting for a train :P

jackpoz replied:
It always confuses me to see PRs opened by other people than the author of the code, I thought tkrokli wrote all the changes

tkrokli replied:
Hmm. Point taken. Would it have made any difference if I had put "by
Riztazz" in caps + bold in the PR description & commit msg?

On Sat, Sep 10, 2016, 23:51 jackpoz <notifications@github.com> wrote:

> In src/server/scripts/EasternKingdoms/ShadowfangKeep/shadowfang_keep.cpp
> <https://github.com/TrinityCore/TrinityCore/pull/17922#discussion_r78284436>
> :
>
> > +            while (auto eventId = events.ExecuteEvent())
> > +            {
> > +                switch (eventId)
> > +                {
> > +                    case EVENT_ARUGAL_CURSE:
> > +                        if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 1, 30.0f, true))
> > +                            DoCast(target, SPELL_ARUGAL_CURSE);
> > +
> > +                        events.ScheduleEvent(eventId, Seconds(15));
> > +                        break;
> > +                    case EVENT_ARUGAL_TELEPORT:
> > +                    {
> > +                        auto spell = arugalTeleportSpells.front();
> > +                        DoCast(spell);
> > +                        // rotate array, can be done with some kind of shuffle. Up to you to figure out what is best.
> > +                        std::rotate(arugalTeleportSpells.begin(), arugalTeleportSpells.begin() + 1, arugalTeleportSpells.end());
>
> It always confuses me to see PRs opened by other people than the author of
> the code, I thought tkrokli wrote all the changes
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/TrinityCore/TrinityCore/pull/17922/files/7ce72f0775af7a3e5786410e65de373478ebf689#r78284436>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AErSsKaRswJd5k4-4xJVKjumVAP_WOT-ks5qoyZYgaJpZM4J5U3J>
> .
>

cooler-SAI commented:
why not merge? all works good!

jackpoz commented:
@cooler-SAI providing feedback to PRs helps understand if it works or if there might be some open issue, something like "I have been testing this PR in last week and got no player reports about it" will surely speed things up, just like your "all works good".

wrathix commented:
was awsome to test this with my gaming community, great work on Arugal

SnapperRy committed:
Update shadowfang_keep.cpp

SnapperRy commented:
Hope I didn't break anything. Arrays aren't my thing.

SnapperRy committed:
Whops.

SnapperRy committed:
Remove whitespace.

Aokromes commented:
what's missing here?

tkrokli commented:
Good question. I think the few issues I talked about in #trinity before opening this PR have been fixed, but
I will do my best to test this again today (solo). It would be really nice if someone could test this with a couple of friends, just to observe that it works as intended and does not need more fixing.

Keader commented:
looking only c++ part, looks ready for me

tkrokli commented:
Should I change `'author'` to be Riztazz instead of me for my main commits? To be effective, I guess it would need me to merge the PR, although there may be some other way in git that I don't know yet.

cooler-SAI commented:
I tested with Friends - all works GOOD!!

Riztazz committed:
Core/Scripts: Archmage Arugal event script

Implement actual boss script for Archmage Arugal
- add yells used in boss event to DB
- use proper spells instead of only melee fighting
- add DataType in shadowfang_keep.h
- event scheduling in the script

by Riztazz, closes issue #17916

Riztazz commented on src/server/scripts/EasternKingdoms/ShadowfangKeep/shadowfang_keep.cpp:
```diff
@@ -207,124 +206,124 @@ class npc_arugal_voidwalker : public CreatureScript
-                        DoCastVictim(SPELL_VOIDBOLT);
-                        events.ScheduleEvent(eventId, Seconds(5));
-                        break;
+
```
whats the point of that if?

tkrokli replied:
Will remove it and push again.
Anything else I should remove too?

tkrokli replied:
Should I remove the `return;` line too?

Riztazz committed:
Remove added merge lines

Redundant lines added during local merge,
removing those to correct the script.

tkrokli commented:
@SnapperRy , @Treeston : please take a quick look to check that I haven't messed up your changes. :-)

Riztazz committed:
Remove another unwanted change

Don't know how those lines got in there,
or even where they came from.

Commit:
SQL rename (update)
